File: app/services/tenants.py
"""TenantService: ciclo de vida de tenants, usuários e memberships (pontos 19, 8).

Criar um tenant cria também seu primeiro usuário OWNER. A api_key do tenant é a
credencial de consumo/master; usuários autenticam pela própria api_key e o papel
vem da membership (owner/member).
"""

import logging
import secrets

from app import rag
from app.config import DEFAULT_TENANT_ID
from app.db import ensure_default_tenant
from app.domain import Membership, Tenant, User
from app.errors import ConflictError, NotFoundError, ValidationError
from app.repositories import (
    AgentRepository,
    MembershipRepository,
    TenantRepository,
    UserRepository,
)
from app.textutil import slugify

log = logging.getLogger("blip-agent.services.tenants")


def _new_key(prefix: str) -> str:
    return f"blip-{prefix}-{secrets.token_urlsafe(24)}"


class TenantService:
    def __init__(self):
        self._tenants = TenantRepository()
        self._users = UserRepository()
        self._members = MembershipRepository()
        self._agents = AgentRepository()

    # --- tenants ----------------------------------------------------------- #
    def list(self) -> list[Tenant]:
        return self._tenants.list()

    def get(self, tenant_id: str) -> Tenant:
        tenant = self._tenants.get(tenant_id)
        if tenant is None:
            raise NotFoundError(f"Tenant '{tenant_id}' não encontrado.")
        return tenant

    def create(self, data: dict) -> tuple[Tenant, User]:
        """Cria o tenant + o primeiro usuário owner. Retorna (tenant, owner)."""
        tenant_id = slugify(data.get("id") or data["name"], fallback="tenant")
        if self._tenants.exists(tenant_id):
            raise ConflictError(f"Já existe um tenant com id '{tenant_id}'.")
        self._tenants.insert(tenant_id, data["name"], _new_key("tenant"))
        owner = self._ensure_user(data["owner_email"], data.get("owner_name", ""))
        self._members.upsert(tenant_id, owner.id, "owner")
        log.info("Tenant criado: %s (owner=%s)", tenant_id, owner.email)
        return self._tenants.get(tenant_id), owner

    def delete(self, tenant_id: str) -> dict:
        """Exclui o tenant. Agentes/produtos/memberships saem por cascade (FK);
        os dados RAG de cada agente são removidos explicitamente (banco à parte)."""
        if not self._tenants.exists(tenant_id):
            raise NotFoundError(f"Tenant '{tenant_id}' não encontrado.")
        chunks = sum(rag.delete_agent_data(a.id)
                     for a in self._agents.list_for_tenant(tenant_id))
        self._tenants.delete(tenant_id)
        return {"deleted": tenant_id, "deleted_chunks": chunks}

    # --- membros ----------------------------------------------------------- #
    def list_members(self, tenant_id: str) -> list[dict]:
        out = []
        for m in self._members.list_for_tenant(tenant_id):
            user = self._users.get(m.user_id)
            out.append({"user_id": m.user_id, "role": m.role,
                        "email": user.email if user else "", "name": user.name if user else ""})
        return out

    def add_member(self, tenant_id: str, email: str, role: str, name: str = "") -> dict:
        if role not in ("owner", "member"):
            raise ValidationError("Papel inválido (use 'owner' ou 'member').")
        user = self._ensure_user(email, name)
        self._members.upsert(tenant_id, user.id, role)
        log.info("Membership %s -> %s (%s)", email, tenant_id, role)
        return {"user_id": user.id, "email": user.email, "role": role,
                "api_key": user.api_key}

    def remove_member(self, tenant_id: str, user_id: str) -> None:
        membership = self._members.get(tenant_id, user_id)
        if membership is None:
            raise NotFoundError("Membership não encontrada.")
        if membership.role == "owner" and self._members.count_owners(tenant_id) <= 1:
            raise ValidationError("Não é possível remover o último owner do tenant.")
        self._members.delete(tenant_id, user_id)

    def membership_of(self, tenant_id: str, user_id: str) -> Membership | None:
        return self._members.get(tenant_id, user_id)

    # --- bootstrap --------------------------------------------------------- #
    def ensure_default_tenant(self) -> Tenant:
        ensure_default_tenant()
        return self._tenants.get(DEFAULT_TENANT_ID)

    # --- internos ---------------------------------------------------------- #
    def _ensure_user(self, email: str, name: str) -> User:
        existing = self._users.get_by_email(email)
        if existing:
            return existing
        user_id = f"usr-{secrets.token_urlsafe(8)}"
        self._users.insert(user_id, email, name, _new_key("user"))
        return self._users.get(user_id)

File: src/server/services/tenants.ts
/**
 * TenantService (porta `app/services/tenants.py`): ciclo de vida de tenants,
 * usuários e memberships. Criar um tenant cria também seu primeiro usuário OWNER.
 */

import { randomBytes } from "node:crypto";

import { DEFAULT_TENANT_ID } from "../config";
import type { Membership, MemberRole, Tenant, User } from "../domain";
import { ConflictError, NotFoundError, ValidationError } from "../errors";
import { getLogger } from "../logging";
import * as rag from "../rag";
import { AgentRepository } from "../repositories/agents";
import { MembershipRepository, TenantRepository, UserRepository } from "../repositories/tenants";
import { slugify } from "../textutil";

const log = getLogger("blip-agent.services.tenants");

function urlsafeToken(bytes: number): string {
  return randomBytes(bytes).toString("base64url");
}
function newKey(prefix: string): string {
  return `blip-${prefix}-${urlsafeToken(24)}`;
}

export interface TenantCreateServiceData {
  id: string | null;
  name: string;
  ownerEmail: string;
  ownerName: string;
}

export interface MemberView {
  userId: string;
  role: string;
  email: string;
  name: string;
}

export interface MemberCreatedView {
  userId: string;
  email: string;
  role: string;
  apiKey: string;
}

export class TenantService {
  // Repositórios injetados por construtor (com default) — mesmo padrão de
  // AgentService/AnalyticsService: testável, sem estado global.
  constructor(
    private readonly tenants = new TenantRepository(),
    private readonly users = new UserRepository(),
    private readonly members = new MembershipRepository(),
    private readonly agents = new AgentRepository(),
  ) {}

  // --- tenants ------------------------------------------------------------- //
  list(): Promise<Tenant[]> {
    return this.tenants.list();
  }

  async get(tenantId: string): Promise<Tenant> {
    const tenant = await this.tenants.get(tenantId);
    if (tenant === null) throw new NotFoundError(`Tenant '${tenantId}' não encontrado.`);
    return tenant;
  }

  /** Cria o tenant + o primeiro usuário owner. Retorna { tenant, owner }. */
  async create(data: TenantCreateServiceData): Promise<{ tenant: Tenant; owner: User }> {
    const tenantId = slugify(data.id || data.name, "tenant");
    if (await this.tenants.exists(tenantId)) {
      throw new ConflictError(`Já existe um tenant com id '${tenantId}'.`);
    }
    await this.tenants.insert(tenantId, data.name, newKey("tenant"));
    const owner = await this.ensureUser(data.ownerEmail, data.ownerName);
    await this.members.upsert(tenantId, owner.id, "owner");
    log.info(`Tenant criado: ${tenantId} (owner=${owner.email})`);
    return { tenant: (await this.tenants.get(tenantId))!, owner };
  }

  /** Exclui o tenant. Agentes/produtos/memberships saem por cascade (FK); o RAG
   * de cada agente é removido explicitamente. */
  async delete(tenantId: string): Promise<{ deleted: string; deleted_chunks: number }> {
    if (!(await this.tenants.exists(tenantId))) {
      throw new NotFoundError(`Tenant '${tenantId}' não encontrado.`);
    }
    const tenantAgents = await this.agents.listForTenant(tenantId);
    let chunks = 0;
    for (const a of tenantAgents) chunks += await rag.deleteAgentData(a.id);
    await this.tenants.delete(tenantId);
    return { deleted: tenantId, deleted_chunks: chunks };
  }

  // --- membros ------------------------------------------------------------- //
  async listMembers(tenantId: string): Promise<MemberView[]> {
    const out: MemberView[] = [];
    for (const m of await this.members.listForTenant(tenantId)) {
      const user = await this.users.get(m.userId);
      out.push({ userId: m.userId, role: m.role, email: user?.email ?? "", name: user?.name ?? "" });
    }
    return out;
  }

  async addMember(tenantId: string, email: string, role: MemberRole, name = ""): Promise<MemberCreatedView> {
    if (role !== "owner" && role !== "member") {
      throw new ValidationError("Papel inválido (use 'owner' ou 'member').");
    }
    const user = await this.ensureUser(email, name);
    await this.members.upsert(tenantId, user.id, role);
    log.info(`Membership ${email} -> ${tenantId} (${role})`);
    return { userId: user.id, email: user.email, role, apiKey: user.apiKey };
  }

  async removeMember(tenantId: string, userId: string): Promise<void> {
    const membership = await this.members.get(tenantId, userId);
    if (membership === null) throw new NotFoundError("Membership não encontrada.");
    if (membership.role === "owner" && (await this.members.countOwners(tenantId)) <= 1) {
      throw new ValidationError("Não é possível remover o último owner do tenant.");
    }
    await this.members.delete(tenantId, userId);
  }

  membershipOf(tenantId: string, userId: string): Promise<Membership | null> {
    return this.members.get(tenantId, userId);
  }

  // --- bootstrap ----------------------------------------------------------- //
  /** Garante o tenant `default` (idempotente). Loga a api_key uma vez. */
  async ensureDefaultTenant(): Promise<Tenant> {
    const existing = await this.tenants.get(DEFAULT_TENANT_ID);
    if (existing) return existing;
    const apiKey = newKey("tenant");
    await this.tenants.insert(DEFAULT_TENANT_ID, "Default", apiKey);
    log.warn(`Tenant '${DEFAULT_TENANT_ID}' criado. api_key=${apiKey} (guarde — usada para chat/consumo).`);
    return (await this.tenants.get(DEFAULT_TENANT_ID))!;
  }

  // --- internos ------------------------------------------------------------ //
  private async ensureUser(email: string, name: string): Promise<User> {
    const existing = await this.users.getByEmail(email);
    if (existing) return existing;
    const userId = `usr-${urlsafeToken(8)}`;
    await this.users.insert(userId, email, name, newKey("user"));
    return (await this.users.get(userId))!;
  }
}
